### PR TITLE
ssh: Use SFTP when possible

### LINKED ISSUE #830: ssh: Support SFTP at least for push/pull/fetch/status
Currently we require full ssh access in order to be able to perform `copy` operation that is needed for external deps/outs/cache scenarios. But for simple push/pull/fetch/status we don't need that operation and can use strictly SFTP.

### PR DESCRIPTION
**Pull request by mroutis**
Close #830

### CODE REVIEW

**Comment by mroutis** [comment]
:thinking: it looks like Python 2.7 doesn't support star `*` unpacking

File: dvc/remote/ssh.py [hunk 1]
```diff
@@ -284,107 +305,86 @@ def download(
...
+                    continue
 
     def list_cache_paths(self):
-        ssh = self.ssh(host=self.host, user=self.user, port=self.port)
-        cmd = "find {} -type f -follow -print".format(self.prefix)
-        stdout = self._exec(ssh, cmd)
-        flist = stdout.split()
-        ssh.close()
+        with self.ssh(**self.path_info) as ssh:
+            cmd = "find {} -type f -follow -print".format(self.prefix)
```
Comment (efiop) [inline dvc/remote/ssh.py]: list_cache_paths is used for `dvc status -c`, so we also need to re-implement this function without using `find` command.

File: dvc/remote/ssh.py [hunk 2]
```diff
@@ -184,68 +194,88 @@ def _exec(self, ssh, cmd):
...
+        dest = to_info["path"]
+        dest_dir = posixpath.dirname(src)
 
-        dname = posixpath.dirname(to_info["path"])
-        self._exec(s, "mkdir -p {}".format(dname))
-        self._exec(s, "cp {} {}".format(from_info["path"], to_info["path"]))
-
-        if not ssh:
-            s.close()
+        with self.ssh(**from_info) as ssh:
```
Comment (efiop) [inline dvc/remote/ssh.py]: path_info might contain additional entries that are not related to the ssh connection(e.g. schema, which is one of the reasons why tests fail).

File: tests/test_data_cloud.py [hunk 3]
```diff
@@ -672,8 +672,6 @@ def _prepare_repo(self):
         self.main(
             ["remote", "add", "-d", TEST_REMOTE, self.cloud._cloud.cache_dir]
         )
-        self.create(self.DATA, self.DATA_CONTENTS)
```
Comment (mroutis) [inline tests/test_data_cloud.py]: Already added in `tests/basic_env.py::setUp`

**Comment by efiop** [comment]
> @efiop, it should be working with this last patch

@mroutis CI still fails though 🙁

**Comment by mroutis** [comment]
> @mroutis CI still fails though slightly_frowning_face
Ready, @efiop, sorry for the false alarm :sweat_smile: ; I'll wait until the test suite finishes the next time
